modules/custom/dkan_datastore/src/Plugin/QueueWorker/DatastoreImportQueue.php

```diff
@@ -0,0 +1,211 @@
+    }
+  }
+
+  /**
+   * 
+   * Sanitise input data for item processing.
+   * 
+   * @param array $data
+   * @return array
+   */
+  protected function sanitizeData(array $data): array {
+
+    if (!isset($data['uuid'], $data['resource_id'], $data['file_path'])) {
+      throw new SuspendQueueException('Queue input data is invalid. Missing required `uuid` or `resource_id`, `file_path`');
+    }
+
+    $data['import_config']     = $data['import_config'] ?? [];
+    $data['file_is_temporary'] = $data['file_is_temporary'] ?? FALSE;
+
+    // State of process.
+    $data['queue_iteration']   = $data['queue_iteration'] ?? 0;
+    $data['rows_done']         = $data['rows_done'] ?? 0;
+    $data['import_fail_count'] = $data['import_fail_count'] ?? 0;
+
+    return $data;
+  }
+
+  /**
+   * Update and validate the state of the queue on success/pause.
+   *
+   * @param array $data The state of the queue.
+   * @param IManager $manager Import manager
+   * @return array Data with updated state info.
+   * @throws SuspendQueueException If the state is invalid.
+   */
+  protected function refreshQueueState(array $data, IManager $manager): array {
+    // Update the state as it were.
+    $newRowsDone = $manager->numberOfRecordsImported();
+
+    // Try to detect if import is stalled.
+    // it shouldn't go backwards but just in case..
+    if ($newRowsDone - $data['rows_done'] <= 0) {
+      $data['import_fail_count'] ++;
+      $this->log(RfcLogLevel::WARNING, "Import for {$data['uuid']} seemd to be lagging behind {$data['import_fail_count']} times. Rows done:{$data['rows_done']} vs {$newRowsDone}");
+    }
+
+    // Suspend further processing.
+    if ($data['import_fail_count'] > static::STALL_LIMIT) {
+      $this->log(RfcLogLevel::ERROR, "Import for {$data['uuid']} lagged for {$data['import_fail_count']} times. Suspending.");
+      throw new SuspendQueueException("Import for {$data['uuid']}[{$data['file_path']}] appears to have stalled past allowed limits.");
+    }
+
+    // otherwise we can keep going.
+    $data['queue_iteration']++;
+    $data['rows_done'] = $newRowsDone;
+
+    return $data;
+  }
+
+  /**
+   * Perfoms cleanup based on iput data.
+   *
+   * @param array $data
+   */
+  protected function cleanup(array $data) {
+    if ($data['file_is_temporary']) {
+      \Drupal::service('file_system')
+        ->unlink($data['file_path']);
+    }
+  }
+
+  /**
+   *
+   * @param mixed $uuid
+   * @param mixed $resourceId
+   * @param string $filePath
+   * @param array $importConfig
+   * @return \Dkan\Datastore\Manager\IManager
+   */
+  protected function getManager($uuid, $resourceId, string $filePath, array $importConfig) {
+    /** @var \Drupal\dkan_datastore\Manager\DatastoreManagerBuilder $managerBuilder */
+    $managerBuilder = \Drupal::service('dkan_datastore.manager.datastore_manager_builder');
+
+    /** @var \Dkan\Datastore\Manager\IManager $manager */
+    $manager = $managerBuilder->setResource($resourceId, $filePath)
```

Comment:
@ketwaroo I believe that once you set the resource, the UUID is ignored by the build process, but we still have to pass that information just to ignore it. Maybe we should have two methods to set the resource, the one that exists and a second one that takes a UUID and sets it in that way. The resource is required, so maybe we just throw an exception if someone calls build without setting a resource previously.

Reply:
Ah. yes. I see what you mean. I'll amend shortly

In queue mode, the uuid is passed mainly so we have something to name the remporary file and to record it in the logs for easier lookup.